### PR TITLE
prometheus-emitter

### PR DESCRIPTION
**Pull request by michaelschiff**
Adds a new extension `prometheus-emitter` to expose Druid metrics for collection directly by a Prometheus server.

This PR has:
- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/incubator-druid/blob/master/licenses.yaml)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

##### Key changed/added classes in this PR
 * `org.apache.druid.emitter.prometheus.*`

### CODE REVIEW

**Comment by michaelschiff** [comment]
Opened now to get any comments on approach before finishing enumerating the different metrics

File: extensions-contrib/prometheus-emitter/src/main/java/org/apache/druid/emitter/prometheus/PrometheusEmitterConfig.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  @JsonProperty
+  private final String path;
+
+  @JsonCreator
+  public PrometheusEmitterConfig(
+      @JsonProperty("namespace") @Nullable String namespace,
+      @JsonProperty("path") @Nullable String path
+  )
+  {
+    this.namespace = namespace != null ? namespace : "druid";
```
Comment (jinfwhuang) [inline extensions-contrib/prometheus-emitter/src/main/java/org/apache/druid/emitter/prometheus/PrometheusEmitterConfig.java]: The default here is slightly hidden. Would this be better if set the default in the field itself, something like `@JsonProperty("namespace") @Nullable String namespace = "druid"` .

Comment (jinfwhuang) [inline extensions-contrib/prometheus-emitter/src/main/java/org/apache/druid/emitter/prometheus/PrometheusEmitterConfig.java]: The default here is slightly hidden. Would this be better if set the default in the field itself, something like `private final String namespace = "druid"` .

Comment (michaelschiff) [reply extensions-contrib/prometheus-emitter/src/main/java/org/apache/druid/emitter/prometheus/PrometheusEmitterConfig.java]: I don't have a strong opinion about this.  For what its worth the `x != null ? x : default` pattern is how the other EmitterConfigs handle null valued options.

Comment (jinfwhuang) [reply extensions-contrib/prometheus-emitter/src/main/java/org/apache/druid/emitter/prometheus/PrometheusEmitterConfig.java]: Okay. I don't have a super strong opinion either. I think the other way around with inline string at the point of declaration is clearer; however, if the existing pattern is inline string in assignment, that is fine too.

**Comment by michaelschiff** [comment]
CC @xvrl @gianm for review

**Comment by michaelschiff** [comment]
Screen Shot 2019-11-01 at 9 35 49 AM
Things are working well in our development clusters.  Please let me know if there is anything else you'd like to see before going forward with review.

**Comment by jihoonson** [comment]
@michaelschiff apologize for the delayed review. I'll take a look the PR soon. I'm closing and reopen the PR to trigger the CI (Druid repository name has changed due to the deincubation and the CI failure logs have gone).

**Comment by michaelschiff** [comment]
@jihoonson Sure - let me know if you need anything from me

File: extensions-contrib/prometheus-emitter/src/test/java/org/apache/druid/emitter/prometheus/PrometheusEmitterTest.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package org.apache.druid.emitter.prometheus;
```
Comment (jihoonson) [inline extensions-contrib/prometheus-emitter/src/test/java/org/apache/druid/emitter/prometheus/PrometheusEmitterTest.java]: The license is missing in this file.

**Comment by jihoonson** [comment]
Oops, @michaelschiff sorry, I forgot about this PR. Would you please resolve the conflicts? Also, Git says this PR changes 808 files, but are they all yours? Please double check this PR contains only your changes.

**Comment by cxlRay** [comment]
clone this emitter and used in my test cluster
I cant find ingest metric in prometheus, is that a bug?

**Comment by michaelschiff** [comment]
@cxlRay do you see the appropriate port open by each of the druid processes? If so can you try to `curl` the metrics endpoint there?

**Comment by michaelschiff** [comment]
@jihoonson FYI @Tiaaa will be taking over this PR (I've changed jobs and can't commit to the source fork). I'll help however I can to see this over the line though.

**Comment by Tiaaa** [comment]
@michaelschiff @jihoonson Travis build succeeded now.

**Comment by jihoonson** [comment]
@michaelschiff thanks for the update. @Tiaaa would you please double check that this PR contains only your changes?

**Comment by jihoonson** [comment]
@Tiaaa thank you for the update. I will take a look through the PR soon. Would you mind updating the version of the extension to 0.19.0-SNAPSHOT?

**Comment by cxlRay** [comment]
@michaelschiff thans for your answer, curl the port what I see like this
`HELP druid_ingest_events_processed Number of events successfully processed per emission period.
TYPE druid_ingest_events_processed counter`

but this metric has value in my log file  
`2020-03-31 16:52:31 INFO [MonitorScheduler-0] org.apache.druid.emitter.prometheus.PrometheusEmitter - druid reporter to prometheus [ingest/events/processed], [druid/middleManager],[6138]`

**Comment by michaelschiff** [comment]
@cxlRay That's the expected response format from the prometheus servlet (meaning it has at least started and registered some of the expected metrics). If you are not seeing druid metrics in prometheus double check that your prometheus server is scraping these hosts (they should be present in `up`)

**Comment by cxlRay** [comment]
@michaelschiff  when peon task start, I got the error `2020-03-31 19:12:46 ERROR [main] org.apache.druid.emitter.prometheus.PrometheusEmitter - Unable to start prometheus HTTPServer
java.net.BindException: Address already in use`,this is because of the prometheus port has started when middlemanager start

I am not sure if it is caused by this
What shoud I do if like this

**Comment by michaelschiff** [comment]
@cxlRay If you are running multiple druid processes on the same physical machine you would need to configure them to use a different port.  The port used by prometheus is set in the runtime.properties of each service.  The default is the same for all of them (which is fine for cases where there is only single service per machine)

**Comment by cxlRay** [comment]
@michaelschiff Please list how to set port used of peon, What I confused is the service of peon task is created dynamic

**Comment by michaelschiff** [comment]
@cxlRay you're right - the cluster we were testing this in does not use the indexing service.  When additional Peon JVMs are started there will be a port conflict.  @Tiaaa we'll need to address this before we can merge

**Comment by busecolak** [comment]
Hello, what is the status of prometheus-emitter? Is there an estimate on when it will be enabled?

**Comment by Tiaaa** [comment]
@busecolak The changes from this PR verified working on coordinator, historical, brokers and indexer. Peon task prometheus emitter we haven't test yet. (Sorry for lost track on this for so long, will pick this up and finish the testing)

**Comment by exherb** [comment]
looks like defaultDimensions.json is out of date?

**Comment by Tiaaa** [comment]
@exherb 
Removed 'query/intervalChunk/time' since it's being deprecated. Any other particular metric you find deprecated?

**Comment by michaelschiff** [comment]
@jihoonson @Tiaaa closed again due to inactivity :( 

I have been using this extension in my clusters for the last 6 months without issue (Without a push-gateway, so not collecting indexing task metrics yet).  I can set this up and start validating this collection path too if that's blocking getting this merged.  @Tiaaa can you bring this PR up to date?  @jihoonson What do you think about getting this merged after that?

**Comment by Tiaaa** [comment]
@michaelschiff re-opened it here https://github.com/apache/druid/pull/10412 cherry-picking all relevant commits.

**Comment by TheRum** [comment]
Do we have any Milestone/Timeline to get this emitter released ?

**Comment by Tiaaa** [comment]
@TheRum currently is in pending review state, hopefully get it in around end of October/mid-November.

**Pull request closed without merging**